### PR TITLE
kernel: workq: Fix type errors in delayable work handlers

### LINKED ISSUE #41942: k_delayable_work being used as k_work in work's handler
The `k_delayable_work` being passed into the work handler as `struct k_work *work` is passed into `CONTAINER_OF` directly without the `k_work_delayable_from_work` to derive the address to the container.

I imagine that the worst case scenario being the code modifies some unintended memory addresses and causes corruptions, but I'm not sure, since some of them have been in the upstream for a few years and don't seem to cause issues.

See https://github.com/ycsin/zephyr/commit/47c0dd506549bbff8bc67b5ff37288ac4faefd60, this commit is done by searching for `CONTAINER_OF(work`, and manually checking each of them, so there might be human error and if the handler k_work arg isn't named as the conventional `work`, then it wont appear in my search result.

### PR DESCRIPTION
**Pull request by ycsin**
The `struct k_work *work` argument of the work handler of a delayable_work should be converted using `k_work_delayable_from_work` before passing into `CONTAINER_OF` to derive the address of its parent struct.

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
Networking part LGTM

**Comment by ycsin** [comment]
DNM, pending reviews from code owners

**Review by trond-snekvik — APPROVED** [review]

**Comment by andyross** [comment]
I know this is probably something "everyone knows", but it would good to call out in the commit message that the reason all this code isn't broken is that the work field is the first element of the delayable, so the pointers have the same value already.  That is, explain this as a typesafety fix and not a logic bug.  This would be how I'd do it, FWIW:

```
kernel: workq: Fix type errors in delayable work handlers

A common pattern here was to take the work item as the subfield of a
containing object.  But the contained field is not a k_work, it's a
k_work_delayable.

Things were working only because the work field was first, so the
pointers had the same value.  Do things right and fix things to
produce correct code if/when that field ever moves within delayable.
```

**Comment by ycsin** [comment]
> This would be how I'd do it, FWIW..

I agree with you and decided to just use your suggestion (shamelessly), I hope you don't mind.

**Review by ceolin — APPROVED** [review]

**Comment by cfriedt** [comment]
Thanks for this @ycsin - just out of curiosity, did you run a particular static analysis tool to pinpoint the areas that needed to be fixed? If so, which tool?

**Comment by ycsin** [comment]
> Thanks for this @ycsin - just out of curiosity, did you run a particular static analysis tool to pinpoint the areas that needed to be fixed? If so, which tool?

I did it manually #41942

> this commit is done by searching for CONTAINER_OF(work, and manually checking each of them, so there might be human error and if the handler k_work arg isn't named as the conventional work, then it wont appear in my search result.

**Comment by cfriedt** [comment]
@ycsin - you're going to hate me for this, but I'm guessing a lot of this also applies to v2.7, so I'm applying the backport label 😬

**Pull request merged**
